lighthouse-core/lib/client-rect-functions.js

```diff
@@ -0,0 +1,311 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/**
+ * @param {LH.Artifacts.ClientRect} cr
+ * @param {{x:number, y:number}} point
+ */
+/* istanbul ignore next */
+function rectContainsPoint(cr, {x, y}) {
+  return cr.left <= x && cr.right >= x && cr.top <= y && cr.bottom >= y;
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect} cr1
+ * @param {LH.Artifacts.ClientRect} cr2
+ */
+/* istanbul ignore next */
+function rectContains(cr1, cr2) {
+  const topLeft = {
+    x: cr2.left,
+    y: cr2.top,
+  };
+  const topRight = {
+    x: cr2.right,
+    y: cr2.top,
+  };
+  const bottomLeft = {
+    x: cr2.left,
+    y: cr2.bottom,
+  };
+  const bottomRight = {
+    x: cr2.right,
+    y: cr2.bottom,
+  };
+  return (
+    rectContainsPoint(cr1, topLeft) &&
+    rectContainsPoint(cr1, topRight) &&
+    rectContainsPoint(cr1, bottomLeft) &&
+    rectContainsPoint(cr1, bottomRight)
+  );
+}
+
+/**
+ * Merge client rects together. This may result in a larger overall size than that of the individual client rects.
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ */
+function simplifyClientRects(clientRects) {
+  clientRects = filterOutTinyClientRects(clientRects);
+  clientRects = filterOutClientRectsContainedByOthers(clientRects);
+  clientRects = mergeTouchingClientRects(clientRects);
+  return clientRects;
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function filterOutTinyClientRects(clientRects) {
+  // 1x1px rect shouldn't be reason to treat the rect as something the user should tap on.
+  // Often they're made invisble in some obscure way anyway, and only exist for e.g. accessibiliity.
+  const nonTinyClientRects = clientRects.filter(
+    rect => rect.width > 1 && rect.height > 1
+  );
+  if (nonTinyClientRects.length > 0) {
+    return nonTinyClientRects;
+  }
+  return clientRects;
+}
+
+const rectContainsString = `
+  ${rectContainsPoint.toString()}
+  ${rectContains.toString()};
+`;
+
+/**
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function filterOutClientRectsContainedByOthers(clientRects) {
+  const filteredOutRects = new Set();
+  return clientRects.filter(cr => {
```

Comment:
nit: the filter but also keeping state with a bool return has me a bit confused, maybe just refactor this to a normal double `for` loop so we can just return the set at the end?

Reply:
We could keep track of the rects we want to keep instead of the ones we want to filter out, but then we'd need to infer what we decided to filter out based on that. I played around with some options, but I'm not quite sure what you have in mind.

Agreed that the current code doesn't read well.

Reply:
I was thinking the flip to rects to keep too

wouldn't the check just then be `!rectsToKeep.has(possiblyContainingRect)`? or did I miss a step?


is the below the same?
```js
for (const cr of clientRects) {
  for (const possiblyContainingRect of clientRects) {
    if (possiblyContainingRect === cr) continue
    if (!rectsToKeep.has(possiblyContainingRect)) continue
    if (rectContains()) { rectsToKeep.delete(cr); break; }
  }
}
```

Reply:
Oh, in my mind I was adding to rectsToKeep instead of deleting from it. This is much more sensible 👍